File: src/main/java/jp/osak/haggledehaghag/repository/RuleAccessRepository.java
package jp.osak.haggledehaghag.repository;

import jp.osak.haggledehaghag.model.RuleAccess;
import org.springframework.data.jdbc.repository.query.Query;
import org.springframework.data.repository.CrudRepository;

import java.util.List;
import java.util.Optional;

public interface RuleAccessRepository extends CrudRepository<RuleAccess, RuleAccess> {
    @Query("""
        SELECT ra.*
        FROM rule_access as ra
        JOIN rule as r ON ra.rule_id = r.id
        WHERE r.game_id = :gameId AND player_id = :playerId
    """)
    List<RuleAccess> findAllByGameIdAndPlayerId(int gameId, int playerId);

    Optional<RuleAccess> findByRuleIdAndPlayerId(int ruleId, int playerId);
}

File: src/main/java/jp/osak/haggledehaghag/repository/RuleAccessRepository.kt
package jp.osak.haggledehaghag.repository

import jp.osak.haggledehaghag.model.RuleAccess
import org.springframework.data.jdbc.repository.query.Modifying
import org.springframework.data.jdbc.repository.query.Query
import org.springframework.data.repository.CrudRepository

interface RuleAccessRepository : CrudRepository<RuleAccess, Int> {
    @Query(
        """
        SELECT ra.*
        FROM rule_access as ra
        JOIN rule as r ON ra.rule_id = r.id
        WHERE r.game_id = :gameId
    """
    )
    fun findAllByGameId(gameId: Int): List<RuleAccess>

    fun findAllByPlayerId(playerId: Int): List<RuleAccess>

    fun findByRuleIdAndPlayerId(ruleId: Int, playerId: Int): RuleAccess?

    @Modifying
    @Query("DELETE FROM rule_access WHERE rule_id = :ruleId")
    fun deleteAllByRuleId(ruleId: Int)

    @Modifying
    @Query("DELETE FROM rule_access WHERE rule_id = :ruleId AND type = :accessType")
    fun deleteAllByRuleIdAndType(ruleId: Int, accessType: RuleAccess.Type)
}
